### ISSUE TITLE
Performance issue with ARES_FLAG_STAYOPEN and PTR lookups with lots of aliases?

### ISSUE THREAD

**Issue by timwoj** [c-ares/c-ares#452]
This is a weird one that I'm not entirely convinced isn't just me doing something wrong, but I thought I'd open the issue and hope someone can tell me. I have an application that makes PTR queries using `ares_create_query()` and `ares_send()`. It's been working fine for a while, but recently I made a request for an IP that returned a large number of host aliases in the response. I've reduced it down to the attached C file. I'm building it with `gcc -lcares -Og -g cares_apple.c` and running under gnu time with `gtime -f "%e" ./a.out`.

It takes about 10 seconds to return on my system. If I comment out setting `ARES_FLAG_STAYOPEN` it returns in under a second. The IP I'm looking up there is one of Apple's and it returns a list of ~20 hostnames. If I change to an IP that I know only returns on hostname, like 1.1.1.1, it returns in under a second whether the flag is set or not.

The delay appears to waiting on `select` in `wait_ares`, but that doesn't explain why it would only delay like that for the one IP and only with the flag set.

Reproducing code:
[cares_apple.c.txt](https://github.com/c-ares/c-ares/files/7974989/cares_apple.c.txt)

**Comment by bradh352** [comment]
Interesting.  My guess is c-ares is doing something bad with STAYOPEN if a query needs to switch from udp to tcp due to the result size being too large.  I doubt may projects use STAYOPEN since most DNS queries occur over UDP and are by definition connectionless so there's not much of a hit for closing the file descriptor when the query count hits zero.

It may be a few weeks before I can take a deeper dive on this issue as I'm swamped at the moment.

**Comment by gvanem** [comment]
AFAICS, your `cares_apple.c` program is basically the same as:
`adig s 1.1.1.1 -f stayopen -t ptr 10.144.253.17.in-addr.arpa`

This takes 10 sec on Windows too.

**Comment by timwoj** [comment]
> I doubt may projects use STAYOPEN since most DNS queries occur over UDP and are by definition connectionless so there's not much of a hit for closing the file descriptor when the query count hits zero.

We use it because we have an internal event loop, and DNS responses are just part of that loop. Having to wake up the loop every time we make a query because the socket's file descriptor changes and we need to insert the new one isn't super ideal.

**Comment by bradh352** [comment]
You should be registering socket callbacks instead to be notified when descriptors you need to monitor change.  Your solution sounds very fragile if you can't handle the file descriptors changing as they definitely will when going from udp to tcp, or when the first dns server doesn't respond and it needs to go onto the next.

I'm not saying there isn't a bug in c-ares by any means, just you need to also make a few changes.

**Comment by timwoj** [comment]
> You should be registering socket callbacks instead to be notified when descriptors you need to monitor change. Your solution sounds very fragile if you can't handle the file descriptors changing as they definitely will when going from udp to tcp, or when the first dns server doesn't respond and it needs to go onto the next.

Right, I am. Whenever I get a socket callback I have to wake up our event loop to add the new file descriptor to the kqueue, resulting in extra churn in the loop. Maybe it's not a big deal in the end, but I'd have to run some tests. My intention in using `ARES_FLAG_STAYOPEN` was to hold onto a file descriptor as long as I could and avoid that extra churn if possible.

**Comment by dopheide-esnet** [comment]
Another data point, in testing the sample on my end, the delay is 30s instead of 10s.

**Comment by bradh352** [comment]
Pretty sure this may be related to #266.  I recently hit a bug with a test case I was writing in relation to multiple queries, so its on my todo list, it specifically impacted TCP not UDP.

**Comment by bradh352** [comment]
looks like i may need to dig into this, as even with my PR to fix #266 this seems to still happen
```
time ./build/bin/adig -s 1.1.1.1 -f stayopen -t ptr 10.144.253.17.in-addr.arpa
id: 7116
flags: qr rd ra 
opcode: QUERY
rcode: NOERROR
Questions:
	10.144.253.17.in-addr.arpa.		PTR
Answers:
	10.144.253.17.in-addr.arpa.	194	PTR	www.brkgls.com.
	10.144.253.17.in-addr.arpa.	194	PTR	asia.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.ca.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.co.uk.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.de.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.es.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.fr.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.it.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.nl.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.ai.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.au.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.bo.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.cn.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.co.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.do.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.gy.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.hn.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.lk.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.mx.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.my.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.pa.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.pe.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.py.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.sg.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.tt.
	10.144.253.17.in-addr.arpa.	194	PTR	apple.com.uy.
	10.144.253.17.in-addr.arpa.	194	PTR	guide.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	shake.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	brkgls.com.
	10.144.253.17.in-addr.arpa.	194	PTR	icloud.com.
	10.144.253.17.in-addr.arpa.	194	PTR	iphone.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	podcast.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	firewire.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	livepage.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	seminars.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	applejava.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	world-any.aaplimg.com.
	10.144.253.17.in-addr.arpa.	194	PTR	advertising.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	applescript.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	itunespartner.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	iworktrialbuy.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	aperturetrialbuy.apple.com.
	10.144.253.17.in-addr.arpa.	194	PTR	squeakytoytrainingcamp.com.
	10.144.253.17.in-addr.arpa.	194	PTR	appstore.com.
NS records:
Additional records:

real	0m10.052s
user	0m0.002s
sys	0m0.006s
```

**Comment by bradh352** [comment]
Upon further investigation, I believe the issue is the method to determine if all queries are done.  Its using ares_fds() to determine if the loop should end, this is wrong.   Especially if the connections are staying open, it could last indefinitely.  In this case, its getting closed by the remote after 10s or so, and then ends.

Instead, the result of ares_timeout() should be used.  If it returns NULL, then there are no outstanding queries.

**Comment by bradh352** [comment]
See PR #452 for how to solve it in your own code.  The PR will resolve it for the adig tool.
